CRT Switchres Capabilities

alphanu1 opened the pull request:
## Guidelines

1. Rebase before opening a pull request
2. If you are sending several unrelated fixes or features, use a branch and a separate pull request for each
3. If possible try squashing everything in a single commit. This is particularly beneficial in the case of feature merges since it allows easy bisecting when a problem arises


## Description

[Description of the pull request, detail any issues you are fixing or any features you are implementing]

## Related Issues

[Any issues this pull request may be addressing]

## Related Pull Requests

[Any other PRs from related repositories that might be needed for this pull request to work]

## Reviewers

[If possible @mention all the people that should review your pull request]

fr500 commented on configuration.c:
```diff
@@ -1485,6 +1486,7 @@ static struct config_uint_setting *populate_settings_uint(settings_t *settings,
 #endif
    SETTING_UINT("audio_out_rate",               &settings->uints.audio_out_rate, true, out_rate, false);
    SETTING_UINT("custom_viewport_width",        &settings->video_viewport_custom.width, false, 0 /* TODO */, false);
+   SETTING_UINT("CRT_Switch_Resolution_super",  &settings->uints.CRT_Switch_Resolution_super,      true, CRT_Switch_Resolution_super, false);//ben
```
please don't use uppercase for variables, it will trigger every dev's OCD

bparker06 replied:
Also @alphanu1 no C99 comments are allowed in the codebase as we still have to support C89-only platforms.

alphanu1 replied:
No worries i'll run though and change things.

fr500 commented on menu/menu_setting.c:
```diff
@@ -3327,6 +3327,34 @@ static bool setting_append_list(
                   SD_FLAG_NONE);
+				  
+				  
+			CONFIG_BOOL( //ben
```
please follow the indentation style from the rest of the code.
We use 3 spaces for tabs.

bparker06 commented on gfx/video_crt_switch.c:
```diff
@@ -0,0 +1,218 @@
+
+void switch_res(int width, int height, int f_restore){  // windows function to swith resolutions
+	
+	DEVMODE curDevmode;
```
All of this is Windows-specific and probably Windows-version-specific (e.g. Win95/NT may not support all kinds of live mode switching), regardless it should not be in a generic file like this and broken out to OS-specific backends.

bparker06 commented on Makefile.common:
```diff
@@ -981,7 +981,8 @@ ifeq ($(HAVE_PARPORT), 1)
 ifneq ($(findstring Win32,$(OS)),)
-   OBJ += input/drivers/winraw_input.o
+   OBJ += input/drivers/winraw_input.o \
+		  gfx/video_crt_switch.o 
```
I think not every version of Windows may support all manners of mode switching (or RAWINPUT for that matter)... and MSVC builds do not use this file at all, instead the source file must be added in griffin/griffin.c.

alphanu1 replied:
This switching method works on all windows versions from Windows XP - Windows 10 as long as CRTEmudriver is installed

alphanu1 replied:
Raw input was already here

bparker06 commented on gfx/video_crt_switch.c:
```diff
@@ -0,0 +1,218 @@
+
+			// This new setting satisfies all the criteria. Try it.
+			devmode.dmFields |= DM_PELSWIDTH | DM_PELSHEIGHT | DM_BITSPERPEL | DM_DISPLAYFREQUENCY;
+			LONG res = ChangeDisplaySettings(&devmode, CDS_TEST);
```
Variables not declared at the top of a function (before any other function calls or non-declarating assignments) breaks C89.

bparker06 commented on intl/msg_hash_us.c:
```diff
@@ -888,6 +888,13 @@ int menu_hash_get_help_us_enum(enum msg_hash_enums msg, char *s, size_t len)
             }
             break;
+			
+		case MENU_ENUM_LABEL_CRT_SWITCH_RESOLUTION: snprintf(s, len, "SET CRT"); //ben
```
Make sure the case label is on its own line separate from any other code.

bparker06 commented on intl/msg_hash_lbl.h:
```diff
@@ -1085,6 +1085,10 @@ MSG_HASH(MENU_ENUM_LABEL_USE_THIS_DIRECTORY,
       "use_this_directory")
 MSG_HASH(MENU_ENUM_LABEL_VIDEO_ALLOW_ROTATE,
       "video_allow_rotate")
+MSG_HASH(MENU_ENUM_LABEL_CRT_SWITCH_RESOLUTION, 
```
Variables can't have spaces in them.

bparker06 commented on README.md:
```diff
@@ -144,3 +144,115 @@ To configure joypads, use the built-in menu or the `retroarch-joyconfig` command
 
 Instructions for compiling and installing RetroArch can be found in the [Libretro/RetroArch Documentation Center](https://docs.libretro.com/).
+
+## CRT SwitchRes
```
Is `SwitchRes` some existing name being used? @twinaphex should we instead be using a generic term?

alphanu1 commented on README.md:
```diff
@@ -144,3 +144,115 @@ To configure joypads, use the built-in menu or the `retroarch-joyconfig` command
+
+CRT Switch res will turn on, on the fly. However, you will need to restart retroarch to disable it. With CRT SwitchRes enable Retroarch will start in 2560 @ 480.  
+
+Before enabling the CRT SwitchRes options pleasae make sure you have installed CRTEmudriver and installed some modelines. The minimun modelins for all games to switch correctly are:
```
CRTEmudriver is needed for windows. it installs a driver allowing it to switch to low resolutions. this is not need for Linux.  I have not implemented  code for Linux switch, however, the code base is already there. just the the final resolution switch for Linux yet.

alphanu1 commented on README.md:
```diff
@@ -144,3 +144,115 @@ To configure joypads, use the built-in menu or the `retroarch-joyconfig` command
+
+## Super Resolutions
+
+The default super resolution is 2560. It is displayed just under the CRT switch option, which can be found in video settings. This can be changed within the retroarch.cfg. The only compatible resolutions are 1920, 2560 and 3840. Any other resolutions will be ignored and native swithing will be activated.
```
Super resolutions are high resolutions on the width, x scale. They are designed so that most if not all width resolutions divide to an exact integer. This makes it be scales of the fly to fit within this resolution without artifacts.

Should this be explained in the Read me?

alphanu1 commented on README.md:
```diff
@@ -144,3 +144,115 @@ To configure joypads, use the built-in menu or the `retroarch-joyconfig` command
+
+## Native Resolutions
+
+On the if any other number is used for super resolutions the native it activeted. if native is what you want then you will need a whole new set of modelines:
```
This has been edited

alphanu1 commented on README.md:
```diff
@@ -144,3 +144,115 @@ To configure joypads, use the built-in menu or the `retroarch-joyconfig` command
+- 2560 x 512 @ 50.000000
+- 2560 x 576 @ 50.000000
+
+Idealy install all these modelines and everything will work great.
```
Edited

alphanu1 commented:
looks like I have a few changes to make. I'll sort them out soon and update.

alphanu1 commented:
I have changes all the code to the request above.

hizzlekizzle commented:
@twinaphex do you have anything to add? any concerns?

twinaphex commented:
@alphanu1 

There are still comments like this in configuration.h -

'//ben'

Let's take these comments out entirely.

twinaphex commented:
https://github.com/alphanu1/MME4CRT/blob/master/gfx/video_crt_switch.h#L21

Why do we have static variables in a header? I think it would be a very bad idea to start doing this in the codebase.

alphanu1 commented:
@twinaphex I have made these changes. All should be good to go.

twinaphex commented:
OK, if within 5 to 6 hours there are no objections remaining, I will merge this in its current state.

bparker06 commented on gfx/video_crt_switch.c:
```diff
@@ -0,0 +1,165 @@
+#include <string.h>
+#include <stdlib.h>
+
+#include <gfx/video_driver.h>
```
This is invalid (however gcc doesn't complain) and fails completely on MSVC with an error, use `#include "video_driver.h"` instead, and these CRT files (the OS-specific ones) need to go in their own directory.

alphanu1 replied:
I thought I had already done this, it turns out I did not.  This is done

bparker06 commented on gfx/video_crt_win.c:
```diff
@@ -0,0 +1,153 @@
+
+
+			devmode.dmFields |= DM_PELSWIDTH | DM_PELSHEIGHT | DM_BITSPERPEL | DM_DISPLAYFREQUENCY;
+			LONG res = ChangeDisplaySettings(&devmode, CDS_TEST);
```
This breaks C89 still and won't compile on MSVC at all (at least 2010 and older).

alphanu1 replied:
What compile errors are you getting here? devmode is used in windows to communicate with all devices from printers to even a mouse. 

It should compile on anything back to 2003 or even earlier.

bparker06 commented:
Also @alphanu1 could you please write a page for this on docs.libretro.com? From a support perspective, we at least need to be able to point people to something that explains all the prerequisites and how this works in general. Thanks

twinaphex commented:
Yeah, it'd really help with dealing with a bunch of user questions if a document gets written that explains how to use this feature.

The code issues that @bparker06 raised should be fixed, after they are fixed we can then start moving towards merging this.

alphanu1 commented:
Once I have changed the necessary issues, I will write up a document for the docs.libretro

alphanu1 commented:
There are other methods for generating and installing modlines for Windows. However, CRTemudriver is by far the easiest but yes it is limited to old Nvidia cards but works with most ATI cards.

Linux does not need anything extra installed to work but I'm still in the process of implementing this.

@bparker06 where in griffin.c do the files need to be implemented? I can have a look through but it will be quicker with a guide.

alphanu1 commented:
Also @bparter06 CRTemudriver does work on all versions of Windows from XP up to 10

alphanu1 commented on gfx/video_driver.c:
```diff
@@ -1414,13 +1419,16 @@ void video_driver_monitor_adjust_system_rates(void)
 
-   timing_skew = fabs(1.0f - info->fps / video_refresh_rate);
-
+  if (crt_switching_active  == true){
```
Sorted

alphanu1 commented on gfx/video_driver.c:
```diff
@@ -1414,13 +1419,16 @@ void video_driver_monitor_adjust_system_rates(void)
-
+  if (crt_switching_active  == true){
+	timing_skew = fabs(1.0f - info->fps / video_driver_core_hz);
+  }else {
```
Changed all of these

alphanu1 commented on gfx/video_crt_switch.h:
```diff
@@ -0,0 +1,26 @@
+void switch_res_crt(int width, int height);
+void aspect_ratio_switch(int width,int height);
+void switch_crt_hz();
+void video_restore();
```
This was  in a couple of files. Done

alphanu1 commented on gfx/video_crt_linux.h:
```diff
@@ -0,0 +1,21 @@
+ *  You should have received a copy of the GNU General Public License along with RetroArch.
+ *  If not, see <http://www.gnu.org/licenses/>.
+*/
+void check_first_run();
```
Done

alphanu1 commented on gfx/video_crt_switch.h:
```diff
@@ -0,0 +1,26 @@
+void check_first_run();
+void screen_setup_aspect(int width, int height);
+void switch_res_crt(int width, int height);
+void aspect_ratio_switch(int width,int height);
```
Done

alphanu1 commented:
I've made all the changes except for griffin.c. I need to run some test to make sure nothing has broken. So not all the files have been updated on Github yet. I will upload once I'm sure all is OK. @twinaphex @bparker06 

Can anyone help me out with griffin.c?

bparker06 commented:
Regarding griffin.c, you would just `#include` the source C files directly in a relevant section of the file, for example here: https://github.com/libretro/RetroArch/blob/master/griffin/griffin.c#L801

you might put `#include "../gfx/video_crt_switch.c"`

and then maybe here: https://github.com/libretro/RetroArch/blob/master/griffin/griffin.c#L189

you could put `#include "../gfx/video_crt_win.c"`

alphanu1 commented:
Ok, I have made all the changes and removed the need for OS specific files. Please let me know if you find anything else wrong, so I can get this finished.

I will now start on writing up a doc for people to follow.

@twinaphex  @bparker06

alphanu1 commented:
I'll also carry on the work for Linux too.

twinaphex commented:
OK @bparker06 is everything alright now or are there still issues remaining?

bparker06 commented:
Still lots and lots of random style issues, and there's still tabs in a few files. The interface API I mentioned [here](https://github.com/libretro/RetroArch/pull/6469#pullrequestreview-108248984) doesn't seem to be implemented, I don't see a PR for a docs page, and was the linux support supposed to be in here before it is merged?

twinaphex commented:
OK, so here is what I can do to pour some water with the wine -

I can offer to clean up the random style issues.

The only thing I'd still like to see from you then @alphanu1 would be that docs page.

@bparker06 Can you make more concrete how that interface API should be done? Maybe @alphanu1 is not certain of what you are requesting here. What about the Linux stuff? Has it been made concrete what that should amount to?

alphanu1 commented:
About the API, I have removed the need for different files dependent on OS. There is one function within the video_crt_switch.c which is used to switch resolutions, it is in here I am using #if define to detect the operating system being compiled for, this intern will run the correct switching method for the specified OS.  

In all honesty I thought I had removed all tabs. I went through the code and deleted them from  everything I clanged or added and replaced them with 3 spaces. I have obviously missed a couple. Sorry about that.

I'm still working on linux. It is functional but only on my system currently, I need to add detection for the connected monitor. This is needed to add the new mode-line to switch to, as all systems are different with many outputs there is no point in guessing they all would be connected to VGA-0, as this would most likely fail. An alternative would be for a user to create a .xprofole with the relevant mode-lines and correct output ID.

twinaphex commented:
OK, so here is the plan - I am going to merge this and clean up afterwards in terms of code styling issues.

If the interface API is still missing, I can consult with @bparker06 and get that done after merging. I understand the concerns about portability and having driver/OS-agnostic code, but according to @hunterk " I don't think there's any way to avoid it", so we have to kinda go with what we have here.

@alphanu1 @hunterk With this being merged, we now need to work on the documentation side. Hopefully you guys can lead that to its successful conclusion. Can you guys focus on that now?

hizzlekizzle commented:
@alphanu1 I made a wiki post for documentation at the forum, assuming you don't already have anything written yet:
https://forums.libretro.com/t/15-khz-crt-documentation-wiki/15671

I figured you and/or the people using it can explain the workflow and settings required to get it going and then we'll move that content over to the documentation site.

alphanu1 commented:
Sounds good. I will get on this soon.

alphanu1 commented:
@twinaphex  You now have this going through your resolution changing back-end. However, does this mean it should be able to switch Linux with your back end.

twinaphex commented:
This is the implementation for Windows.

https://github.com/libretro/RetroArch/blob/master/gfx/display_servers/dispserv_win32.c#L282

You could add a function callback implementation to the X11 Linux driver (inside the same directory - dispserv_x11.c) and implement it that way. Note - there is still one WIN32 ifdef remaining - the system metric calls - I am still undecided as to how we will abstract away that, so don't worry about that for now - if you need to, you can still put a linux ifdef there.

And yes, this display server driver is automatically used at runtime based on whether or not an X11 server is available and whether or not we are on Linux/Unix.
